---
title: "Exercise 2.3"
author: "Sere Williams"
date: "2/13/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
library(tidyverse)
library(dplyr)
```


## Exercise 2.3 from Modern Statistics for Modern Biologists

A sequence of three nucleotides codes for one amino acid. There are 4 nucleotides, thus 4^3 would allow for 64 different amino acids, however there are only 20 amino acids requiring only 20 combinations + 1 for an end signal. The code is redundant. Here we use the tuberculosis genome to understand codon bias. 

# a) Explore the data mtb using table to tabulate the AmAcid and Codon variables.

Each amino acid is encoded by 1-6 tri-nucleotide combinations. 
```{r}
mtb = read.table("../data/M_tuberculosis.txt", header = TRUE)
codon_no <- rowSums(table(mtb))
codon_no
```
The PerThousands of each codon can be visualized, where each bar represents a different codon. But what does the "PerThousands" variable mean?
```{r, fig.width=4, fig.height=3}
ggplot(mtb, aes(x=Codon, y=PerThous)) +
  geom_col()+
  facet_wrap(~AmAcid, scales="free") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

# b) How was the PerThous variable created?

The sum of all of the numbers of codons gives you the total number of codons in the M. tuberculosis genome: `all_codons`. Remember that this is not the size of the M. tuberculosis genome, but the number of codons in all M. tuberculosis genes. To get the size of the genome you'd have to multiply each codon by 3 (for each nucleotide) and add non-coding nucleotides (which we do not know from this dataset).

```{r}
all_codons = sum(mtb$Number)
all_codons
```

The PerThousands variable is derived by dividing the number of occurances of the codon of interest by the total number of codons. Because this number is then small and hard to intrepret, multipling it by 1000 gives a value that is easy to make sense of quickly. Here is an example for proline. The four values returned align to the four codons that each code for proline. 
```{r}
pro  =  mtb[ mtb$AmAcid == "Pro", "Number"]
pro/all_codons*1000
```

# c) Write an R function that you can apply to the table to find which of the amino acids shows the strongest codon bias, i.e., the strongest departure from uniform distribution among its possible spellings.

First, let's look at the expected frequencies of each codon. 
```{r}
codon_expected <- data.frame(codon_no) %>%
  rownames_to_column() %>%
  rename(`AmAcid` = rowname) %>%
  mutate(prob_codon = 1/codon_no)
codon_expected
```

Next, calculate the observed frequencies for each codon seen in the dataset:
```{r}
codon_observed <- full_join(mtb, codon_expected, by = "AmAcid") %>%
  group_by(AmAcid) %>%
  mutate(am_length = sum(Number))%>%
  mutate(prob_observed = Number / am_length) %>%
  select(-codon_no, -am_length) %>%
  mutate(expected_PerThous = Number*prob_codon)
codon_observed
```

Using the chi-squared distribution, a chi-squared test statistic can be used to show codon bias. 

```{r}
chi_stat <- data.frame(chi_squared = ((codon_observed$Number - codon_observed$expected_PerThous)^2)/(codon_observed$expected_PerThous))
chi_stat

#I'm getting an error here. I'm not sure what is wrong.
codon_observed <- codon_observed() %>%
  mutate(chi-squared = ((codon_observed$Number - codon_observed$expected_PerThous)^2)/(codon_observed$expected_PerThous)) %>% 
  select(-prob_codon, -prob_observed)
codon_observed

#Also an error here. 
codon_observed <- codon_observed() %>%
  mutate(chi-squared = chi_stat$chi_squared) %>% 
  select(-prob_codon, -prob_observed)
codon_observed


#Additional ideas from students in class. I'm not able to run it, but I like the idea of using a function.  
stat = function(codon_expected, exptd = 20 * pvec) {
   sum((codon_df - exptd)^2 / exptd)
}

stat

B = 1000
simulstat = replicate(B, {
  randomtab10 = sapply(stat, function(s) { rmultinom(1, s, p0) })
  stat(randomtab10, expectedtab10)
})
S1 = stat(tab10, expectedtab10)
sum(simulstat >= S1)
```


