---
title: Exercise solution for Chapter 2, Part 1
author: Sere Williams
date: '2020-02-19'
slug: exercise-solution-for-chapter-2
draft: yes
categories:
  - exercises
  - Chapter 2
tags:
  - exercises
  - Chapter 2
subtitle: ''
summary: ''
authors: [sere-williams]
lastmod: '2020-02-19T15:25:48-07:00'
featured: no
image:
  caption: ''
  focal_point: ''
  preview_only: no
projects: []
---

```{r setup, include=FALSE, echo = TRUE}
knitr::opts_chunk$set(echo = TRUE)
```

As always, load libraries first.
```{r message = FALSE}
library(ggplot2)
library(tidyverse)
library(dplyr)
```


## Exercise 2.3 from Modern Statistics for Modern Biologists

A sequence of three nucleotides codes for one amino acid. There are 4 nucleotides, thus $4^3$ would allow for 64 different amino acids, however there are only 20 amino acids requiring only 20 combinations + 1 for an "end" signal. (The "start" signal is the codon, ATG, which also codes for the amino acid methionine, so the start signal does not have a separate codon.) The code is redundant. But is the redundancy even among codons that code for the same amino acid? In other words, if alanine is coded by 4 different codons, do these codons code for alanine equally (each 25%), or do some codons appear more often than others? Here we use the tuberculosis genome to explore codon bias. 

# a) Explore the data `mtb` using `table` to tabulate the `AmAcid` and `Codon` variables.

Each amino acid is encoded by 1--6 tri-nucleotide combinations. 

```{r}
mtb = read.table("example_datasets/M_tuberculosis.txt", header = TRUE)
codon_no <- rowSums(table(mtb))
codon_no
```

The `PerThousands` of each codon can be visualized, where each plot represents an amino acid and each bar represents a different codon that codes for that amino acid. But what does the `PerThousands` variable mean?

```{r, fig.height=6, fig.width=8}
ggplot(mtb, aes(x=Codon, y=PerThous)) +
  geom_col()+
  facet_wrap(~AmAcid, scales="free") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

# b) How was the `PerThous` variable created?

The sum of all of the numbers of codons gives you the total number of codons in the M. tuberculosis genome: `all_codons`. Remember that this is not the size of the M. tuberculosis genome, but the number of codons in all M. tuberculosis genes. To get the size of the genome, multiply each codon by 3 (for each nucleotide) and add all non-coding nucleotides (which we do not know from this data set).

```{r}
all_codons = sum(mtb$Number)
all_codons
```

The `PerThousands` variable is derived by dividing the number of occurrences of the codon of interest by the total number of codons. Because this number is small and hard to interpret, multiplying it by 1000 gives a value that is easy to make sense of. Here is an example for proline. The four values returned align to the four codons that each code for proline. 

```{r}
pro  =  mtb[mtb$AmAcid == "Pro", "Number"]
pro / all_codons * 1000
```

# c) Write an R function that you can apply to the table to find which of the amino acids shows the strongest codon bias, i.e., the strongest departure from uniform distribution among its possible spellings.

First, let's look at the expected frequencies of each codon. 

```{r}
codon_expected <- data.frame(codon_no) %>%
  rownames_to_column() %>%
  rename(`AmAcid` = rowname) %>%
  mutate(prob_codon = 1/codon_no)
codon_expected
```

Next, calculate the observed frequencies for each codon seen in the data set and use the chi-squared test statistic to determine if the difference between expected and observed codon frequencies is even or if some codon sequences are used more than others. 

```{r, fig.height=6, fig.width=8}
codon_compared <- mtb %>% 
  group_by(AmAcid) %>% 
  mutate(total = sum(Number),
         n_codons = n(),
         expected = total / n_codons)
codon_compared

ggplot(codon_compared, aes(x=Codon, y=expected)) +
     geom_col()+
     facet_wrap(~AmAcid, scales="free") +
     theme(axis.text.x = element_text(angle = 45, hjust = 1))

codon_compared %>% 
  filter(n_codons > 1) %>% 
  group_by(AmAcid) %>% 
  mutate(chi_squared = ((Number - expected)^2/expected)) %>% 
  summarise(chi_squared = sum(chi_squared),
            n = n()) %>% 
  mutate(p_value = pchisq(chi_squared, df = n-1, log = TRUE, lower.tail = FALSE)) %>% 
  arrange(p_value) %>% 
  mutate(rank = 1:n())
```
As you may notice, these p-values are large, negative (so very tiny) values. In calculating the p-value from chi-squared text statistic, we took the log. If you exponentiate the p-value, you will get the actual p-value, but this number is so small (0.00000000e+00) that it is looses meaning. Instead, we will leave the p-value as a negative and sort the amino acids from most negative to less negative, providing a rank order of amino acids with the most unevenly used codons at 1 and amino acids with more evenly used codons as their rank increases. 
